### PR TITLE
Let Vivarium Engine Profile (including parallel processes)

### PR DESCRIPTION
**Pull request by U8NWXD**
If you set `profile=True` in the `Engine` constructor, then when you call `Engine.end()`, Vivarium will compute profile statistics and store them in `Engine.stats` for you to analyze. The Engine also profiles parallel processes and includes those stats in `Engine.stats`, giving you a full picture of what's taking up runtime in your simulation.

See `vivarium/experiments/test_profile.py` for an example of how to use this.

### CODE REVIEW

File: vivarium/core/engine.py [hunk 1]
```diff
@@ -415,12 +418,19 @@ def __init__(
             progress_bar: shows a progress bar
             emit_config: If True, this will emit the serialized
                 processes, topology, and initial state.
+            profile: Whether to profile the simulation with cProfile.
         """
+        if profile:
+            self.profiler = cProfile.Profile()
+            self.profiler.enable()
```
Comment (eagmon) [inline vivarium/core/engine.py:425]: You need `self.profiler = None` perhaps if `profile=False`

File: vivarium/experiments/test_profiler.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        profile=True,
+    )
+    engine.update(3)
+    engine.end()
+    assert engine.stats is not None
+    stats = engine.stats.strip_dirs()
+    process_a_runtime = stats.stats[  # type: ignore
+        ('test_profiler.py', 16, 'next_update')][3]
+    process_b_runtime = stats.stats[  # type: ignore
+        ('test_profiler.py', 31, 'next_update')][3]
```
Comment (eagmon) [inline vivarium/experiments/test_profiler.py:56]: Out of curiosity, what are those `16`, `31` for?

Comment (U8NWXD) [reply vivarium/experiments/test_profiler.py:56]: Those are the line numbers for where the two `next_update` functions are defined. That's how `cProfile` resolves cases like this where there are 2 functions of the same name in a file

File: vivarium/core/engine.py (hunk 1 above)
Comment (eagmon) [reply vivarium/core/engine.py:426]: looks like you took car of this.

Comment (U8NWXD) [reply vivarium/core/engine.py:426]: I fixed this when fixing some failing tests

**Review by eagmon — APPROVED** [review]
This is great! I'm tempted to remake `profile_runtime.py` with this.

**Comment by U8NWXD** [comment]
Thanks @eagmon! I'll merge it and create a new release

**Pull request merged**